adapters/coap/src/test/java/org/eclipse/hono/adapter/coap/TelemetryResourceTest.java

```diff
@@ -334,6 +337,71 @@ public void testUploadTelemetryWithQoS1() {
                 any());
     }
 
+    /**
+     * TODO.
+     */
+    @Test
+    public void testServerTimeOptionIsSetInResponseWhenRequestedQoS0() {
+        // GIVEN an adapter with a downstream telemetry consumer attached
+        givenAnAdapter(properties);
+        givenATelemetrySenderForAnyTenant();
+        final var resource = givenAResource(adapter);
+
+        // WHEN a device publishes an telemetry message
+        final Buffer payload = Buffer.buffer("some payload");
+        final OptionSet options = new OptionSet();
+        options.setContentFormat(MediaTypeRegistry.TEXT_PLAIN);
+        options.addOption(new Option(0xff20));
+        final CoapExchange coapExchange = newCoapExchange(payload, Type.NON, options);
+        final DeviceUser authenticatedDevice = new DeviceUser("tenant", "device");
+        final CoapContext context = CoapContext.fromRequest(coapExchange, authenticatedDevice, authenticatedDevice, "device", span);
+
+        resource.handlePostRequest(context);
+
+        // THEN the device gets a response indicating success
+        verify(coapExchange).respond(argThat((Response res) -> ResponseCode.CHANGED.equals(res.getCode())));
+        verify(coapExchange).respond(argThat((Response res) -> res.getOptions().hasOption(TIME_OPTION_NUMBER)));
+        verify(coapExchange).respond(argThat((Response res) -> {
+            final byte[] optionTimeValue = res.getOptions().getOtherOption(TIME_OPTION_NUMBER).getValue();
+            final long optionTime = new BigInteger(optionTimeValue).longValue();
+            return System.currentTimeMillis() >= optionTime;
+        }));
+    }
+
+    // TODO: should these tests be somewhere else? This is only testing the telemetry endpoint but these tests are more
+    //       general as they apply to all endpoints. Should this be in CoapContextTest?
+    /**
+     * TODO.
+     */
+    @Test
+    public void testServerTimeOptionIsSetInResponseWhenRequestedQoS1() {
```

Comment:
can we merge this with the previous test case and use `@ParameterizedTest` to reflect the different QoS levels?

Reply:
sure, I can do that.

I was looking into this a bit last night and I'm just battling through understanding how the hono tests are structured. I'll get there eventually.

But I at least added two tests in `CoapContextTest` where I test whether the response includes the time option if 1) the request has the time option or 2) the request has the request parameter set. Those tests work well so I'm wondering whether I should also have these types of tests like I've done here in the `TelemetryResourceTest`. I feel a bit weird about only having it there because it should work for Command and Event resources too.

I was trying to add it to the `AbstractResourceTest` but there I ran into some problems because the `CoapContext` object wasn't being created and manipulated.

So I'm wondering, should I just have the tests that I have now in `CoapContextTest` and call it a day or should I also have tests on the resources themselves? If adding tests to the resources I kinda want to add an `AllResourceTest` class where the test would be parameterized to test all three resource types, CON and NON request types and both test the request option as well as the request parameter way.

Am I overthinking this maybe? ;)

Reply:
I just committed the `CoapContextTest` so you can see what I'm talking about there... the rest is in a bit too much of a WIP status currently for me to commit it.

Reply:
Putting the unit tests into *CoapContextTest* makes sense to me as it is *CoapContext* where the magic happens.
In order to make sure that a device actually gets what it expects, I would suggest to add/amend corresponding integration test cases in `tests/src/test/java/org/eclipse/hono/tests/coap/CoapTestBase`. These tests are being run against both the *telemetry* and the *event* endpoints ...

Reply:
Maybe you want to start digging into these integration tests first. But do not hesitate to ask if you need help with finding your way around. But you already have done a great job in finding the right places to make changes, so I am confident you will succeed here as well :-)